chainer/initializers/__init__.py

```diff
@@ -29,14 +29,14 @@ def generate_array(initializer, shape, xp):
              or :class:`cupy.ndarray` and edits its value.
         shape (tuple): Shape of a return array.
         xp (module): :mod:`cupy` or :mod:`numpy`.
+        dtype: Dtype specifier. If omitted, ``initializer.dtype`` is used.
 
     Returns:
         numpy.ndarray or cupy.ndarray: An initialized array.
 
     """
-    dtype = numpy.float32
-    if hasattr(initializer, 'dtype') and initializer.dtype is not None:
-        dtype = initializer.dtype
+    if dtype is None:
+        dtype = getattr(initializer, 'dtype', numpy.float32)
```

Comment:
How about prefering `initializer.dtype` if both specified?
That's why I suggested `default_dtype` as the name of the argument.

Reply:
Hmm, I feel the argument should provide the way to override other configurations.

Reply:
I think that could be confusing for initializer side.
Initializers with concrete `dtype`s should expect arrays with that `dtype`.

If `initializer.dtype` is just a default dtype, it should have been named as `default_dtype`.

Reply:
Actually, initializer implementations have `if self.dtype is not None: assert array.dtype == self.dtype` in their `__call__` method.

Reply:
Hmmm, ok, I understand the situation, but I'm a little bit against using the name `default_dtype` because it conflicts with the "default dtype" concept introduced in #4510. Another solution is just raising an error when these two (`dtype` and `initializer.dtype`) conflict.

Reply:
That sounds good.